src/helpers/validators/validateString.ts

```diff
@@ -0,0 +1,154 @@
+
+// Import types
+import ValidationResult from './shared/types/ValidationResult';
+
+/*------------------------------------------------------------------------*/
+/* ------------------------------ Constants ----------------------------- */
+/*------------------------------------------------------------------------*/
+
+// define minimum and maximum range for lowercase ASCII chars
+const LOWERCASE_MIN = 65;
+const LOWERCASE_MAX = 90;
+
+// define minimum and maximum range for uppercase ASCII chars
+const UPPERCASE_MIN = 97;
+const UPPERCASE_MAX = 122;
+
+// define minimum and maximum range for ASCII digits
+const DIGIT_MIN = 48;
+const DIGIT_MAX = 57;
+
+/*------------------------------------------------------------------------*/
+/* ---------------------------- Main Function --------------------------- */
+/*------------------------------------------------------------------------*/
+
+/**
+ * Determines whether a given input string is considered valid based on
+ *   the provided requirements.
+ * @author Austen Money
+ * @param input input string
+ * @param opts options for validation
+ * @returns whether input is considered valid according to reqs - if
+ *   valid, returns a cleaned version of input; if invalid, returns
+ *   a string containing error messages describing which requirements
+ *   were not met.
+ */
+const validateString = (
+  input: string,
+  opts: {
+    // whitespace is removed from input before checking and returning
+    ignoreWhitespace?: boolean,
+    // input length must be at least minLen
+    minLen?: number,
+    // input length must be at most maxLen
+    maxLen?: number,
+    // input must only contain letters
+    lettersOnly?: boolean,
+    // input must only contain numbers
+    numbersOnly?: boolean,
+    // input must match given regExp
+    regexTest?: RegExp,
+    // description of regExp test, used to customize error messages
+    regexDescription?: string,
+  },
+): ValidationResult<string> => {
+  // stores all invalid input errors
+  const errorMessages: string[] = [];
+
+  // contains version of input that will be returned
+  let cleanedValue: string = input;
+
+  // remove whitespace if required
+  if (opts.ignoreWhitespace) {
+    cleanedValue = input.replace(/\s+/g, '');
+  }
+
+  // apply max char requirement
+  if (opts.minLen) {
+    if (cleanedValue.length < opts.minLen) {
+      errorMessages.push(INVALID_STRING_ERRORS.MIN_LEN(opts.minLen));
+    }
+  }
+
+  // apply max char requirement
+  if (opts.maxLen) {
+    if (cleanedValue.length > opts.maxLen) {
+      errorMessages.push(INVALID_STRING_ERRORS.MAX_LEN(opts.maxLen));
+    }
+  }
+
+  // apply alphabetical requirement
+  if (opts.lettersOnly) {
+    // remove diacritics
+    cleanedValue = cleanedValue.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
+
+    const containsNonLetters = cleanedValue.split('').some((curr) => {
+      // check that char is an upper or lower-case letter
+      const currCode = curr.charCodeAt(0);
```

Comment:
I think what you're doing here is just fine. The only other idea I have to offer is a multistep process:

```ts
cleanedValue
  // Split into characters
  .split('')
  // Convert into char codes
  .map((char) => {
    return curr.charCodeAt(0);
  })
  // Check for non-letters
  .some((currCode) => {
```

Reply:
I'll leave it up to you to think through which is better

Reply:
I like that! Having every step be its own chained function makes it feel more step-by-step.